source/AI.cpp

```diff
@@ -2764,24 +2764,26 @@ void AI::DoSurveillance(Ship &ship, Command &command, shared_ptr<Ship> &target)
 		}
 
 		unsigned total = targetShips.size() + targetPlanets.size() + targetSystems.size();
+		// If there is nothing for this ship to scan, have it patrol the entire system
+		// instead of drifting or stopping.
 		if(!total)
 		{
-			// If there is nothing for this ship to scan, have it hold still
-			// instead of drifting away from the system center.
-			Stop(ship, command);
+			DoPatrol(ship, command);
 			return;
 		}
-
-		unsigned index = Random::Int(total);
-		if(index < targetShips.size())
-			ship.SetTargetShip(targetShips[index]->shared_from_this());
 		else
 		{
-			index -= targetShips.size();
-			if(index < targetPlanets.size())
-				ship.SetTargetStellar(targetPlanets[index]);
+			unsigned index = Random::Int(total);
+			if(index < targetShips.size())
+				ship.SetTargetShip(targetShips[index]->shared_from_this());
 			else
-				ship.SetTargetSystem(targetSystems[index - targetPlanets.size()]);
+			{
+				index -= targetShips.size();
+				if(index < targetPlanets.size())
+					ship.SetTargetStellar(targetPlanets[index]);
+				else
+					ship.SetTargetSystem(targetSystems[index - targetPlanets.size()]);
+			}
 		}
```

Comment:
This stuff doesn't need to be in an `else` anymore.
Basically, the only change that should be here is that you should call `DoPatrol` instead of `Stop` in the `if` part.
```suggestion
		}

		unsigned index = Random::Int(total);
		if(index < targetShips.size())
			ship.SetTargetShip(targetShips[index]->shared_from_this());
		else
		{
			index -= targetShips.size();
			if(index < targetPlanets.size())
				ship.SetTargetStellar(targetPlanets[index]);
			else
				ship.SetTargetSystem(targetSystems[index - targetPlanets.size()]);
		}
```